### PR TITLE
 Update class code cleaning to allow new omitted assessment class codes

### PR DESCRIPTION
**Pull request by wrridgeway**
The office introduced new class codes this year that start with "OA". Our current syntax for cleaning class codes (`REGEXP_REPLACE(par.class, '([^0-9EXR])', '') AS class`) only preserves the number after 'OA'. We should also stop wiping out the `A` and `B` suffixes for incentive classes.

Some tables have already been updated in a [previous PR](https://github.com/ccao-data/data-architecture/pull/338/files).

### CODE REVIEW

File: aws-athena/views/default-vw_pin_universe.sql
Comment (dfsnow) [inline aws-athena/views/default-vw_pin_universe.sql:1]: **question (blocking):** Is this all the views that have a `class` column where this applies? Did you check the dbt DAG?

Comment (wrridgeway) [reply aws-athena/views/default-vw_pin_universe.sql:1]: This is all the views where the class column was being cleaned. I can expand this see if there are any views that need cleaning *added*, if we want.

File: aws-athena/views/default-vw_pin_condo_char.sql [hunk 1]
```diff
@@ -34,7 +34,7 @@ oby_filtered AS (
     -- We don't include DEACTIVAT IS NULL here since it can disagree with
     -- DEACTIVAT in iasworld.pardat and we'll defer to that table
     WHERE cur = 'Y'
-        AND class IN ('299', '2-99', '399')
+        AND class IN ('299', '399')
```
Comment (wrridgeway) [inline aws-athena/views/default-vw_pin_condo_char.sql:37]: '2-99' is only an issue in pardat.

File: aws-athena/views/default-vw_pin_condo_char.sql [hunk 2]
```diff
@@ -50,7 +50,7 @@ comdat_filtered AS (
...
```
Comment (wrridgeway) [inline aws-athena/views/default-vw_pin_condo_char.sql:53]: '2-99' is only an issue in pardat.

**Comment by wrridgeway** [comment]
## vw_pin_appeal

```
with old as (
select year, count(*) as "old" from default.vw_pin_appeal
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_appeal
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_condo_char

```
with old as (
select year, count(*) as "old" from default.vw_pin_condo_char
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_condo_char
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_exempt

```
with old as (
select year, count(*) as "old" from default.vw_pin_exempt
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_exempt
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_sale

```
with old as (
select year, count(*) as "old" from default.vw_pin_sale
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_sale
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_universe

```
with old as (
select year, count(*) as "old" from default.vw_pin_universe
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_universe
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_shared_input

```
with old as (
select meta_year, count(*) as "old" from model.vw_pin_shared_input
group by meta_year
),
new as (
select meta_year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_model".vw_pin_shared_input
group by meta_year
)

select new.meta_year, new.new, old.old from old
left join new
on new.meta_year = old.meta_year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_ratio_stats

```
with old as (
select year, count(*) as "old" from reporting.vw_ratio_stats
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_reporting".vw_ratio_stats
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_top_5

```
with old as (
select year, count(*) as "old" from reporting.vw_top_5
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_reporting".vw_top_5
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_code_retrieval

```
with old as (
select year, count(*) as "old" from rpie.vw_code_retrieval
group by year
),
new as (
select year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_rpie".vw_code_retrieval
group by year
)

select new.year, new.new, old.old from old
left join new
on new.year = old.year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
## vw_pin_flatfile

```
with old as (
select rpie_year, count(*) as "old" from rpie.vw_pin_flatfile
group by rpie_year
),
new as (
select rpie_year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_rpie".vw_pin_flatfile
group by rpie_year
)

select new.rpie_year, new.new, old.old from old
left join new
on new.rpie_year = old.rpie_year
where  new.new != old.old
```

**Comment by wrridgeway** [comment]
## vw_pin_mailers

```
with old as (
select rpie_year, count(*) as "old" from rpie.vw_pin_mailers
group by rpie_year
),
new as (
select rpie_year, count(*) as "new" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_rpie".vw_pin_mailers
group by rpie_year
)

select new.rpie_year, new.new, old.old from old
left join new
on new.rpie_year = old.rpie_year
where  new.new != old.old
```

no results

**Comment by wrridgeway** [comment]
There's a lot to review here... open to any suggestions as to how to better test the implications of these changes.

File: aws-athena/views/default-vw_pin_condo_char.sql [hunk 3]
```diff
@@ -228,10 +232,7 @@ SELECT DISTINCT
     filled.card,
     filled.lline,
     filled.year,
-    CASE WHEN filled.class = '2-99'
-            THEN '299'
-        ELSE filled.class
-    END AS class,
+    filled.class,
```
Comment (wrridgeway) [inline aws-athena/views/default-vw_pin_condo_char.sql:235]: Should be fixed in the CTE above.

**Comment by wrridgeway** [comment]
```
with old as (
select class, count(*) as "old count" from default.vw_pin_universe
group by class
),
new as (
select class, count(*) as "new count" from "z_ci_341-update-class-code-cleaning-to-allow-new-omitted-assessment-class-codes_default".vw_pin_universe
group by class
)

select new.class as "new class", old.class as "old class", "new count", "old count" from old
full outer join new
on new.class = old.class
```

[complete new vs old class breakout](https://github.com/ccao-data/data-architecture/files/14711561/7c288e7b-e8b8-4a12-a25f-76dbfec96fd2.csv)

File: aws-athena/views/default-vw_pin_condo_char.sql [hunk 4]
```diff
@@ -60,7 +60,9 @@ chars AS (
     SELECT
         par.parid AS pin,
         CASE
```
Comment (dfsnow) [inline aws-athena/views/default-vw_pin_condo_char.sql:62]: **issue (blocking):** For anything like this, where there's a `CASE WHEN` without an `ELSE`, we should add a dbt test for not null.

File: aws-athena/views/default-vw_pin_appeal.sql [hunk 5]
```diff
@@ -1,7 +1,7 @@
 -- View containing appeals by PIN
 SELECT
     htpar.parid AS pin,
-    pardat.class AS class,
+    REGEXP_REPLACE(pardat.class, '[^[:alnum:]]', '') AS class,
```
Comment (dfsnow) [inline aws-athena/views/default-vw_pin_appeal.sql:4]: We need to start manually updating `ccao.class_dict` to ensure that the new classes are included there, otherwise a bunch of existing dbt tests are going to start failing. I opened a separate issue here: https://github.com/ccao-data/data-architecture/issues/362.

File: aws-athena/views/reporting-vw_ratio_stats.sql [hunk 6]
```diff
@@ -6,8 +6,10 @@ WITH classes AS (
     SELECT
         par.parid,
         par.taxyr,
-        par.class,
-        CASE WHEN par.class IN ('299', '399') THEN 'CONDO'
+        REGEXP_REPLACE(par.class, '[^[:alnum:]]', '') AS class,
```
Comment (dfsnow) [inline aws-athena/views/reporting-vw_ratio_stats.sql:9]: **issue:** Likewise on adding a not null test here.

**Review by dfsnow — COMMENTED** [review]
@wrridgeway Need a few more tests here, but otherwise looks okay. We should be _really_ careful with big changes like this and should be on the lookout for any bugs. Lots of this will need an update after #362.

**Comment by dfsnow** [comment]
@wrridgeway What's the status on this PR?

File: aws-athena/views/reporting-vw_ratio_stats.sql (hunk 6 above)
Comment (wrridgeway) [reply aws-athena/views/reporting-vw_ratio_stats.sql:9]: I'll address this [here](https://github.com/ccao-data/data-architecture/pull/367).

**Comment by wrridgeway** [comment]
Added some more dbt tests for columns that can, but shouldn't be null. dbt build is failing because of a bad value in htpar that needs to be fixed, currently.

File: aws-athena/views/default-vw_pin_appeal.sql (hunk 5 above)
Comment (dfsnow) [reply aws-athena/views/default-vw_pin_appeal.sql:4]: @wrridgeway The addition of the `OA` classes here is going to cause test failures to appear in our QC workbooks, since those classes aren't yet defined in the `class_dict` table. We should open a quick PR in lieu of #362 to add `class_dict` as a dbt seed and stub out the OA classes. We can do that as a fast-follow or before merging this PR.

File: aws-athena/views/default-vw_pin_universe.sql
Comment (dfsnow) [inline aws-athena/views/default-vw_pin_universe.sql:1]: Note to future selves that this is just a line ending change

File: aws-athena/views/default-vw_pin_universe.sql [hunk 7]
```diff
@@ -1,146 +1,148 @@
...
+    vwl.econ_coordinated_care_area_num,
+    vwl.econ_coordinated_care_area_data_year,
+    vwl.econ_enterprise_zone_num,
+    vwl.econ_enterprise_zone_data_year,
+    vwl.econ_industrial_growth_zone_num,
+    vwl.econ_industrial_growth_zone_data_year,
+    vwl.econ_qualified_opportunity_zone_num,
+    vwl.econ_qualified_opportunity_zone_data_year,
+    vwl.econ_central_business_district_num,
+    vwl.econ_central_business_district_data_year,
```
Comment (dfsnow) [inline aws-athena/views/default-vw_pin_universe.sql:78]: **question:** Shouldn't this be in a separate PR?

File: dbt/models/default/schema/default.vw_pin_condo_char.yml [hunk 8]
```diff
@@ -89,6 +89,20 @@ models:
             - year
           config:
             error_if: ">70297"
+      - not_null:
+          name: default_vw_pin_condo_char_card_not_null
+          column_name: card
```
Comment (dfsnow) [inline dbt/models/default/schema/default.vw_pin_condo_char.yml:94]: **suggestion:** For simple tests like these, Jean suggested it's probably better to attach them to their respective columns rather than defining them as standalone tests (we get more metadata that way). That said, we aren't exporting them in any meaningful way, so I'm not sure it matters. Cue a longer conversation about unit vs data tests (or exported vs CI tests).

My original goal of defining tests on the default views was to test the _underlying data_, however it seems like almost all of that function has moved to the `test_qc_` tests. Perhaps we should only put data QC tests on sources (iasworld, other ingested stuff) and use other tests as CI checks only. Any thoughts @jeancochrane @wrridgeway?

File: aws-athena/views/default-vw_pin_universe.sql (hunk 7 above)
Comment (wrridgeway) [reply aws-athena/views/default-vw_pin_universe.sql]: I don't know why this is showing up as a change... it's in master, and I don't see a commit on this branch where I changed this?

File: dbt/models/default/schema/default.vw_pin_condo_char.yml (hunk 8 above)
Comment (wrridgeway) [reply dbt/models/default/schema/default.vw_pin_condo_char.yml:94]: Personally, I'm for it. I just had to remove the `card_proration_rate` test for null-ness since the fields it's built on are absolute garbage. Having a view fail every time we try to build it because oby and comdat are poorly maintained doesn't seem like the proper way to address issues in iasworld.

Comment (jeancochrane) [reply dbt/models/default/schema/default.vw_pin_condo_char.yml:94]: Agreed @dfsnow. Once unit tests are released in dbt 1.8 (current target date is 5/9) and we can get started on https://github.com/ccao-data/data-architecture/issues/382 it'll be a lot easier for us to differentiate these two types of tests.

File: aws-athena/views/default-vw_pin_appeal.sql (hunk 5 above)
Comment (wrridgeway) [reply aws-athena/views/default-vw_pin_appeal.sql:4]: Addressed in https://github.com/ccao-data/data-architecture/pull/397.

**Review by dfsnow — APPROVED** [review]
Looks all set to me! Let's merge @wrridgeway
